solr/core/src/java/org/apache/solr/search/PerSegmentViewDocIdSetIterator.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.search;
+
+import java.io.IOException;
+
+import org.apache.lucene.search.DocIdSetIterator;
+
+/**
+ * A {@link org.apache.lucene.search.DocIdSetIterator} which allows single-segment access to a wrapped top-level (multi-segment) DISI.
+ *
+ * Useful as a shim between a top-level data structure, and logic that attempts to operate on it using segment-level docIds.
+ * The same wrapped DISI can be used serially by multiple PerSegmentViewDocIdSetIterators
+ */
+public class PerSegmentViewDocIdSetIterator extends DocIdSetIterator {
```

Comment:
Can you make this an inner class of topLevelDocValuesTermsQuery or anonymous?  If there is a future need to share it _then_ I think we should pull it out.  I doubt it's as generally useful to be a standalone public class.

After further review, I don't think we need the functionality here _at all_... see my future feedback